subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -5,18 +5,176 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#if defined(CONFIG_NET_DEBUG_SOCKETS)
+#define SYS_LOG_DOMAIN "net/tls"
+#define NET_LOG_ENABLED 1
+#endif
+
+#include <stdbool.h>
+
+#include <init.h>
 #include <net/net_context.h>
 #include <net/socket.h>
 
-#include "sockets_tls.h"
+#if defined(CONFIG_MBEDTLS)
+#if !defined(CONFIG_MBEDTLS_CFG_FILE)
+#include "mbedtls/config.h"
+#else
+#include CONFIG_MBEDTLS_CFG_FILE
+#endif /* CONFIG_MBEDTLS_CFG_FILE */
+
+#include <mbedtls/ctr_drbg.h>
+#include <mbedtls/net_sockets.h>
+#include <mbedtls/x509.h>
+#include <mbedtls/x509_crt.h>
+#include <mbedtls/ssl.h>
+#include <mbedtls/error.h>
+#include <mbedtls/debug.h>
+#endif /* CONFIG_MBEDTLS */
+
+/** TLS context information. */
+struct tls_context {
+	/** Information whether TLS context is used. */
+	bool is_used;
+
+	/** Secure protocol version running on TLS context. */
+	enum net_ip_protocol_secure tls_version;
+
+#if defined(CONFIG_MBEDTLS)
+	/** mbedTLS context. */
+	mbedtls_ssl_context ssl;
+
+	/** mbedTLS configuration. */
+	mbedtls_ssl_config config;
+#endif /* CONFIG_MBEDTLS */
+};
+
+/* A global pool of TLS contexts. */
+static struct tls_context tls_contexts[CONFIG_NET_SOCKETS_TLS_MAX_CONTEXTS];
+
+/* A mutex for protecting TLS context allocation. */
+static struct k_mutex context_lock;
+
+/* Initialize TLS internals. */
+static int tls_init(struct device *unused)
+{
+	ARG_UNUSED(unused);
+
+	memset(tls_contexts, 0, sizeof(tls_contexts));
+
+	k_mutex_init(&context_lock);
+
+	return 0;
+}
+
+SYS_INIT(tls_init, APPLICATION, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+
+/* Allocate TLS context. */
+static struct tls_context *tls_alloc(void)
+{
+	int i;
+	struct tls_context *tls = NULL;
+
+	k_mutex_lock(&context_lock, K_FOREVER);
+
+	for (i = 0; i < ARRAY_SIZE(tls_contexts); i++) {
+		if (!tls_contexts[i].is_used) {
+			tls = &tls_contexts[i];
+			memset(tls, 0, sizeof(*tls));
+			tls->is_used = true;
+
+			NET_DBG("Allocated TLS context, %p", tls);
+			break;
+		}
+	}
+
+	k_mutex_unlock(&context_lock);
+
+	if (tls) {
+		mbedtls_ssl_init(&tls->ssl);
+		mbedtls_ssl_config_init(&tls->config);
+	} else {
+		NET_WARN("Failed to allocate TLS context");
+	}
+
+	return tls;
+}
+
+/* Release TLS context. */
+static int tls_release(struct tls_context *tls)
+{
+	if (!PART_OF_ARRAY(tls_contexts, tls)) {
+		NET_ERR("Invalid TLS context");
+		return -EINVAL;
```

Comment:
EBADF is *the* error for such a case, see e.g. http://pubs.opengroup.org/onlinepubs/9699919799/functions/close.html